src/jinja2/environment.py

```diff
@@ -855,7 +949,62 @@ def _load_template(self, name, globals):
         return template
 
     @internalcode
-    def get_template(self, name, parent=None, globals=None):
+    def _check_multi_template_autoescape(
+        self,
+        name: t.Union[None, str, "Template"],
+        parent: t.Optional[str],
+        caller: t.Optional[str],
+    ) -> None:
+        """
+        Raise an Error if we expect different custom escape functions
+        or Markup classes used in a way that a user could get
+        unexpected results.
+
+        :param name: The name of first template if known
+        :param parent: the name or names of the templates that
+        :param caller: the name of the function the
+
+        :raises TemplateConfigurationError
+        """
+        # Currently we check only for extends and mixed auto escape,
+        # so we can exit early
+        if self._allow_mixed_escape_extends or caller != "extends":
+            return
+
+        if isinstance(name, Template):
+            name = None
+        elif not isinstance(name, str):
+            raise ValueError(
+                f"Parameter name is not Template nor string. "
+                f"Given was '{name}' (Type: {type(name)})."
+            )
+        base_class = self.get_markup_class(name)
+        parent_class = self.get_markup_class(parent)
+
+        # The Simplest case it is exactly the same class
+        # We do not check for subclasses because a subclass could have
+        # altered the escape behavior already
+        if base_class == parent_class:
+            return
+
+        # The dynamic nature of get_wrapped_escape_class does not allow
+        # for direct compares, so we need to compare the uid
+        if (
+            hasattr(base_class, "get_class_uid")
+            and hasattr(parent_class, "get_class_uid")
+            and base_class.get_class_uid == parent_class.get_class_uid  # type: ignore
+        ):
+            return
```

Comment:
Why so much complexity? What's wrong with something as simple as `if base_class != parent_class: raise ...`

Reply:
Because I tested it and it won't work. Because the classes given by two calls of get_wrapped_escape_class actually aren't the same even using the lru_cache!

Reply:
ah I didn't notice we're not dealing directly with the custom Markup classes (or escape functions) here..

Reply:
resolved?

Reply:
Ok did some further digging. The lru_cache solves the issue so we don't need the complexity. I removed it, as I could write a test that calls 997.
In my original testing I extracted the Markup classes from the template context that is generated through exec().
That was probably the reason why the lru_cache wasn't working.

Pushing the changes soon.